File: nuido_flow_data/flows/data/create_data_node.py
# THIS FILE IS A PART OF PUBLIC REPOSITORY https://github.com/yonitjio/exploring-odoo
#
# This software is released under the MIT License.
# https://opensource.org/licenses/MIT
#
# THIS SOFTWARE IS EXPERIMENTAL AND FOR EDUCATIONAL PURPOSE ONLY.
# DO NOT USE IT IN PRODUCTION.

import logging

from odoo.tools import safe_eval

from odoo.addons.nuido_flow.flows.core.base_node import BaseNode
from odoo.addons.nuido_flow.flows.tools.tools import get_default_context_for_eval, get_active_record_info

from .tools import get_lookup_nodes

from odoo.addons.nuido_flow.flows.tools.log_const import LOGGER_NAME
_logger = logging.getLogger(LOGGER_NAME)

class CreateDataNode(BaseNode):
    def _process(self, params):
        super()._process(params)

        context = get_default_context_for_eval(self.env)
        if params is not None:
            context['params'] = params

        info = get_active_record_info(self.env)
        context = {**context, **info}

        nodes, lookup_functions = get_lookup_nodes(self)

        context["nodes"] = nodes
        context["lookup"] = lookup_functions

        values = {}
        model = self.definition["model"]
        for field in self.definition["fields"]:
            try:
                value = safe_eval.safe_eval(field["value"], context)
            except:
                _logger.warning("Exception on evaluation: %s", field["value"], exc_info=True)
                value = None

            field = self.env["ir.model.fields"]._get(model, field["name"])
            if field.ttype in ('one2many', 'many2many'):
                if type(value) == list:
                    values.update({ field["name"]: value })
                else:
                    values.update({ field["name"]: [value] })
            else:
                values.update({ field["name"]: value })

        records = self.env[model].create(values)

        return {
            "result": records
        }

File: nuido_flow_data/static/src/components/data/create_data_node.js
// THIS FILE IS A PART OF PUBLIC REPOSITORY https://github.com/yonitjio/exploring-odoo
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT
//
// THIS SOFTWARE IS EXPERIMENTAL AND FOR EDUCATIONAL PURPOSE ONLY.
// DO NOT USE IT IN PRODUCTION.
import { useState } from "@odoo/owl";
import { Node } from "@nuido/components/node";
import { ModelSelectorEx } from "@nuido_flow/components/ui/model_selector_ex";
import { ModelFieldSelectorEx } from "@nuido_flow/components/ui/model_field_selector_ex";
export class CreateDataNode extends Node {
    setup() {
        super.setup();
        this.state = useState({
            model: this.props.node.model,
            modelDescription: this.props.node.model_description,
            fieldName: "",
            fieldValue: ""
        });
    }
    resetFieldState() {
        this.state.fieldName = "";
        this.state.fieldValue = "";
    }
    onModelSelected(model) {
        const { label, technical: value } = model;
        this.props.node.fields = [];
        this.state.model = value;
        this.state.modelDescription = label;
        this.props.node.model = value;
        this.props.node.model_description = label;
    }
    async onModelFieldUpdate(path, info) {
        this.state.fieldName = info.fieldDef ? info.fieldDef.name : "";
        this.refreshEdges();
    }
    onAddField() {
        if (this.state.fieldName !== "" && this.state.fieldValue !== "") {
            this.props.node.addOrUpdateField(this.state.fieldName, this.state.fieldValue);
            this.refreshEdges();
            this.resetFieldState();
        }
    }
    onRemoveField(name) {
        this.props.node.removeRecordMap(name);
        this.refreshEdges();
    }
    get fieldValueId() {
        return `input-${this.props.node.id}}-field-value`;
    }
}
CreateDataNode.template = "nuido_flow_data.create-data-node";
CreateDataNode.components = {
    ...Node.components,
    ModelSelectorEx,
    ModelFieldSelectorEx
};
